core/src/main/java/org/apache/iceberg/util/SortStrategyUtil.java

```diff
@@ -0,0 +1,223 @@
+    }
+    // Then normalize the result
+    return totalScore / (2 * files.size());
+  }
+
+  /**
+   * Removes files that are already sorted (having no overlap on sort column values with other files)
+   * from a collection of file scans.
+   *
+   * @param fileScans The original collection of file scans
+   * @param sortOrder The sort order
+   * @return A collection of file scans after removing sorted ones from the original collection
+   */
+  public static Iterable<FileScanTask> removeSortedFiles(Iterable<FileScanTask> fileScans, SortOrder sortOrder) {
+    List<Integer> sortColIndices = getSortColumnIndices(sortOrder);
+    List<DataFile> dataFiles = FluentIterable.from(fileScans).stream()
+        .map(FileScanTask::file)
+        .collect(Collectors.toList());
+    return FluentIterable.from(fileScans)
+        .filter(f -> hasOverLap(f.file(), dataFiles, sortColIndices, sortOrder.schema()));
+  }
+
+  private static boolean hasOverLap(
+      DataFile file,
+      List<DataFile> dataFiles,
+      List<Integer> sortColIndices,
+      Schema sortSchema) {
+    for (DataFile f : dataFiles) {
+      if (!f.equals(file) && isOverlap(f, file, sortColIndices, sortSchema)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  static List<Integer> getSortColumnIndices(SortOrder sortOrder) {
+    List<Integer> sortColIndices = Lists.newArrayList();
+    for (SortField sortField : sortOrder.fields()) {
+      if (!sortField.transform().isIdentity()) {
+        LOG.info("Cannot measure sortedness score on derived sort column {}",
+            sortField.transform().dedupName());
+        return ImmutableList.of();
+      }
+      sortColIndices.add(sortField.sourceId());
+    }
+    return sortColIndices;
+  }
+
+  // Counts how many data files from a given list have overlaps with a given data point
+  static int overlapCount(
+      List<ByteBuffer> point,
+      List<DataFile> files,
+      List<Integer> sortColIndices,
+      Schema sortSchema) {
+    int overLapCount = 0;
+    for (DataFile file : files) {
+      if (isOverlap(point, file, sortColIndices, sortSchema)) {
+        overLapCount++;
+      }
+    }
+    return overLapCount;
+  }
+
+  private static double overlapScore(
+      List<ByteBuffer> point,
+      List<DataFile> files,
+      List<Integer> sortColIndices,
+      Schema sortSchema) {
+    double overlapScore = overlapCount(point, files, sortColIndices, sortSchema);
+    return 1.0 - (overlapScore - 1.0) / (files.size() - 1.0);
+  }
+
+  // Check if a data point (x[1], x[2],..., x[k]) has overlap with a data file that has lower bounds of
+  // (min[1], min[2],..., min[k]) and upper bounds of (max[1], max[2],..., max[k])
+  // An overlap happens if for j going from 1 to k:
+  // if x[j] < min[j] or x[j] > max[j] then return false
+  // else if min[j <= x[j] <= max[j] and min[j] != max[j] and  then return true
+  // else (meaning min[j] == x[j] == max[j]) continue with j + 1
+  private static boolean isOverlap(
```

Comment:
I think we probably could just compare the pairs outright like we do
https://github.com/apache/iceberg/blob/master/spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteDataFilesAction.java#L1335-L1376

We also probably don't need the manfiest evaluators I think...

Reply:
Also a little confusing that we have "hasOverlap" and "isOverlap", I think we may just want to have a 
```java
boolean overlaps(fileA, fileB, fieldIds) {
   return true if fileA overlaps fileB 
}
```